src/Compilers/CSharp/Portable/Lowering/NullCheckRewriter.cs

```diff
@@ -0,0 +1,68 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.CodeGen;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using System.Collections.Immutable;
+
+namespace Microsoft.CodeAnalysis.CSharp.Lowering
+{
+    internal sealed class NullCheckRewriter : BoundTreeRewriterWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator
+    {
+        private readonly MethodSymbol _method;
+        private readonly SyntheticBoundNodeFactory _fact;
+        private NullCheckRewriter(
+            MethodSymbol method,
+            SyntaxNode syntax,
+            TypeCompilationState compilationState,
+            DiagnosticBag diagnostics)
+        {
+            _method = method;
+            _fact = new SyntheticBoundNodeFactory(method, syntax, compilationState, diagnostics);
+        }
+
+        internal static BoundStatement Rewrite(
+            BoundStatement body,
+            MethodSymbol method,
+            TypeCompilationState compilationState,
+            DiagnosticBag diagnostics)
+        {
```

Comment:
>{ [](start = 8, length = 1)

Consider returning here without any allocations if there are no null-checked parameters:
```C#
if (!method.Parameters.Any(p => p.IsNullChecked))
{
    return body;
}
```

Reply:
Should I get rid of the sawNullChecked var then?

Reply:
Ah, sorry, I'd overlooked the fact that the caller is already checking that at least one parameter is null-checked.

Then either approach sounds reasonable - either checking parameters outside of the rewriter and only calling when there are null-checked parameters, or calling the rewriter always and returning early here. If we stick with the former, then this `method.Parameters.Any(...)` here could be a `Debug.Assert()` instead.

---
In reply to: [295010196](https://github.com/dotnet/roslyn/pull/36471#discussion_r295010196) [](ancestors = 295010196)